src/jdk.compiler/share/classes/com/sun/tools/javac/jvm/ClassReader.java

```diff
@@ -2255,12 +2257,311 @@ public void run() {
                 currentClassFile = classFile;
                 List<Attribute.TypeCompound> newList = deproxyTypeCompoundList(proxies);
                 sym.setTypeAttributes(newList.prependList(sym.getRawTypeAttributes()));
+                addTypeAnnotationsToSymbol(sym, newList);
             } finally {
                 currentClassFile = previousClassFile;
             }
         }
     }
 
+    /**
+     * Rewrites types in the given symbol to include type annotations.
+     *
+     * <p>The list of type annotations includes annotations for all types in the signature of the
+     * symbol. Associating the annotations with the correct type requires interpreting the JVMS
+     * 4.7.20-A target_type to locate the correct type to rewrite, and then interpreting the JVMS
+     * 4.7.20.2 type_path to associate the annotation with the correct contained type.
+     */
+    private static void addTypeAnnotationsToSymbol(
+            Symbol s, List<Attribute.TypeCompound> attributes) {
+        new TypeAnnotationSymbolVisitor(attributes).visit(s, null);
+    }
+
+    private static class TypeAnnotationSymbolVisitor
+            extends Types.DefaultSymbolVisitor<Void, Void> {
+
+        private final List<Attribute.TypeCompound> attributes;
+
+        private TypeAnnotationSymbolVisitor(List<Attribute.TypeCompound> attributes) {
+            this.attributes = attributes;
+        }
+
+        @Override
+        public Void visitClassSymbol(Symbol.ClassSymbol s, Void unused) {
+            ClassType t = (ClassType) s.type;
+            int i = 0;
+            ListBuffer<Type> interfaces = new ListBuffer<>();
+            for (Type itf : t.interfaces_field) {
+                interfaces.add(addTypeAnnotations(itf, classExtends(i++)));
+            }
+            t.interfaces_field = interfaces.toList();
+            t.supertype_field = addTypeAnnotations(t.supertype_field, classExtends(65535));
```

Comment:
can we use a named constant here instead of a literal?

Reply:
That happened as a different change: https://bugs.openjdk.org/browse/JDK-8340024

My understanding was that the best practice for backports was to minimize the number of changes from the original patch, and if a chain of related changes needed to be backported to do them all as separate changes, is that accurate? Would you like me to backport that one as well?

Reply:
@cushon You could fold this change into this one. It makes sense to backport them both. Then use `/issue add JDK-8340024` on this PR.

Reply:
@jerboaa thanks! I will do that.

Do you have thoughts on including #1251 and #1252 in this PR vs. keeping them separate? They're all part of the same logical fix.

Reply:
@cushon I may have been too fast. This is an 21-only fix (the redo) and [JDK-8340024](https://bugs.openjdk.org/browse/JDK-8340024) would be a backport. So to keep this neat work-flow wise it probably makes more sense to remove it again and do it separately (as a backport PR). Sorry for the confusion. this is a pretty special case. I'd also keep the others separate.

Reply:
Ok, sounds good! I've removed that commit, unliked it from that bug, and cancelled the backport request for now.

Reply:
FYI: I've submitted the constant cleanup as a backport here: https://github.com/openjdk/jdk21u-dev/pull/1671